state/presence/pingbatcher.go

```diff
@@ -0,0 +1,267 @@
+		case <-flushTimeout:
+			// TODO (jam): 2017-06-21 I think I was wrong about channel selectivity.
+			// I think the problem was actually that we restart this timer every time we get a ping
+			// Switch back to a central loop and move this timeout
+			// outside of the loop, only creating it when it times out.
+			if err := pb.flush(); err != nil {
+				return errors.Trace(err)
+			}
+			flushTimeout = time.After(pb.nextSleep())
+		case singlePing := <-pb.pingChan:
+			pb.handlePing(singlePing)
+		case flushReq := <-pb.flushChan:
+			// Flush is requested synchronously.
+			// This way we know all pings have been handled. We will
+			// close the channel that was passed to us once we have
+			// finished flushing.
+			if err := pb.flush(); err != nil {
+				return errors.Trace(err)
+			}
+			close(flushReq)
+		}
+	}
+}
+
+// Ping should be called by a Pinger when it is ready to update its time slot.
+// It passes in all of the pre-resolved information (what exact field bit is
+// being set), rather than the higher level "I'm pinging for this Agent".
+// Internally, we synchronize with the main worker loop. Which means that this
+// function will return once the main loop recognizes that we have a ping request
+// but it will not have updated its internal structures, and certainly not the database.
+func (pb *PingBatcher) Ping(modelUUID string, slot int64, fieldKey string, fieldBit uint64) error {
+	ping := singlePing{
+		Slot:      slot,
+		ModelUUID: modelUUID,
+		FieldKey:  fieldKey,
+		FieldBit:  fieldBit,
+	}
+	select {
+	case pb.pingChan <- ping:
+		return nil
+	case <-pb.tomb.Dying():
+		err := pb.tomb.Err()
+		if err == nil {
+			return errors.Errorf("PingBatcher is stopped")
+		}
+		return errors.Trace(err)
+	}
+}
+
+// ForceFlush immediately flushes the current state to the database.
+// This should generally only be called from testing code, everyone else can
+// generally wait the usual wait for updates to be flushed naturally.
+func (pb *PingBatcher) ForceFlush() error {
+	request := make(chan struct{})
+	select {
+	case pb.flushChan <- request:
+		select {
+		case <-request:
+			return nil
+		case <-pb.tomb.Dying():
+			return pb.tomb.Err()
+		}
+	case <-pb.tomb.Dying():
+		return pb.tomb.Err()
+	}
+}
+
+// handlePing is where we actually update our internal structures after we
+// get a ping request.
+func (pb *PingBatcher) handlePing(ping singlePing) {
+	docId := docIDInt64(ping.ModelUUID, ping.Slot)
+	cur, slotExists := pb.pending[docId]
+	if !slotExists {
+		cur.Alive = make(map[string]uint64)
+		cur.Slot = ping.Slot
+		pb.pending[docId] = cur
+	}
+	alive := cur.Alive
+	alive[ping.FieldKey] |= ping.FieldBit
+	pb.pingCount++
+}
+
+// flush pushes the internal state to the database. Note that if the database
+// updates fail, we will still wipe our internal state as it is unsafe to
+// publish the same updates to the same slots.
+// TODO (jam): 2017-06-21 Maybe if we switched from "$inc" to "$bit":
```

Comment:
Performance wise, I don't think it matters, because $inc vs $bit {or} doesn't matter as long as we are keeping the frequency low.
It always bothered me that we're doing bitwise or by using $inc and suffering from "if we ever ping the same slot twice we corrupt the database".
I'll file it as a bug and leave it as tech-debt.